databuilder/databuilder/publisher/neo4j_csv_unwind_publisher.py

```diff
@@ -0,0 +1,471 @@
+            prop_body_update = prop_body_template.render(prop_body_r1=prop_body_r1, prop_body_r2=prop_body_r2)\
+                if self._publish_reverse_relationships else prop_body_r1
+
+        return template.render(START_LABEL=rel_record[RELATION_START_LABEL],
+                               END_LABEL=rel_record[RELATION_END_LABEL],
+                               relationship_stmt=two_way_relationship if self._publish_reverse_relationships
+                               else one_way_relationship,
+                               update_prop_body=prop_body_r1,
+                               prop_body_create=prop_body_create,
+                               prop_body_update=prop_body_update)
+
+    def _create_props_param(self, record_dict: dict) -> dict:
+        params = {}
+        for k, v in record_dict.items():
+            if k.endswith(UNQUOTED_SUFFIX):
+                k = k[:-len(UNQUOTED_SUFFIX)]
+
+            params[k] = v
+        return params
+
+    def _create_props_body(self,
+                           record_dict: dict,
+                           excludes: Set,
+                           identifier: str,
+                           rename_id_to_preserve_ui_data: bool = False) -> str:
+        """
+        Creates properties body with params required for resolving template.
+
+        e.g: Note that node.key3 is not quoted if header has UNQUOTED_SUFFIX.
+        identifier.key1 = 'val1' , identifier.key2 = 'val2', identifier.key3 = val3
+
+        :param record_dict: A dict represents CSV row
+        :param excludes: set of excluded columns that does not need to be in properties
+        (e.g: KEY, LABEL ...)
+        :param identifier: identifier that will be used in CYPHER query as shown on above example
+        :return: Properties body for Cypher statement
+        """
+        # For SET, if the evaluated expression is null, no action is performed. I.e. `SET (null).foo = 5` is a noop.
+        # See https://neo4j.com/docs/cypher-manual/current/clauses/set/
+        if rename_id_to_preserve_ui_data:
+            identifier = \
+                f"(CASE WHEN {identifier}.{PUBLISHED_TAG_PROPERTY_NAME} IS NOT NULL THEN {identifier} ELSE null END)"
+
+        props = []
+        for k, v in record_dict.items():
+            if k in excludes:
+                continue
+
+            if k.endswith(UNQUOTED_SUFFIX):
+                k = k[:-len(UNQUOTED_SUFFIX)]
+
+            props.append(f'{identifier}.{k} = row.{k}')
+
+        if self._add_publisher_metadata:
+            props.append(f"{identifier}.{PUBLISHED_TAG_PROPERTY_NAME} = '{self._publish_tag}'")
+            props.append(f"{identifier}.{LAST_UPDATED_EPOCH_MS} = timestamp()")
+
+        # add additional metadata fields from config
+        for k, v in self._additional_publisher_metadata_fields.items():
+            val = v if isinstance(v, int) or isinstance(v, float) else f"'{v}'"
+            props.append(f"{identifier}.{k} = {val}")
```

Comment:
instead of adhoc handling, can we append this to record_dict and let it handle it.

Reply:
Allison recently added this and I should have pointed it out then.